java_src/foam/dao/index/TreeIndex.java

```diff
@@ -0,0 +1,152 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package foam.dao.index;
+
+import foam.core.FObject;
+import foam.core.PropertyInfo;
+import foam.dao.Sink;
+import foam.mlang.predicate.And;
+import foam.mlang.predicate.Binary;
+import foam.mlang.predicate.Eq;
+import foam.mlang.predicate.False;
+import foam.mlang.predicate.Gt;
+import foam.mlang.predicate.Gte;
+import foam.mlang.predicate.Lt;
+import foam.mlang.predicate.Lte;
+import foam.mlang.predicate.Predicate;
+import foam.mlang.sink.Count;
+import java.util.Comparator;
+
+public class TreeIndex implements Index {
+  protected Index tail;
+  protected PropertyInfo prop;
+  protected long selectCount;
+  protected TreeNode tree;
+  
+  public TreeIndex(PropertyInfo prop) {
+    this.prop = prop;
```

Comment:
The whole body of this method should just be:
  this(prop, ValueIndex.instance());